File: exercises/day20-Architecture-Proposal/java/src/main/java/core/control/ReindeerPowerUnit.java
package core.control;

import external.deer.Reindeer;

public class ReindeerPowerUnit {
    public Reindeer reindeer;
    public MagicPowerAmplifier amplifier;

    public ReindeerPowerUnit(Reindeer reindeer, MagicPowerAmplifier amplifier) {
        this.reindeer = reindeer;
        this.amplifier = amplifier;
    }

    public float harnessMagicPower() {
        if (!reindeer.needsRest()) {
            reindeer.timesHarnessing++;
            return amplifier.amplify(reindeer.getMagicPower());
        }

        return 0;
    }

    public float checkMagicPower() {
        return reindeer.getMagicPower();
    }
}

File: exercises/day20-Architecture-Proposal/kotlin/src/main/kotlin/core/control/ReindeerPowerUnit.kt
package core.control

import external.deer.Reindeer

class ReindeerPowerUnit(val reindeer: Reindeer, val amplifier: MagicPowerAmplifier) {
    fun harnessMagicPower(): Float {
        return if (!reindeer.needsRest()) {
            reindeer.timesHarnessing++
            amplifier.amplify(reindeer.magicPower)
        } else {
            0f
        }
    }

    fun checkMagicPower() = reindeer.magicPower
}
